crypto/thread/internal.c

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright 2019-2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/configuration.h>
+#include <openssl/e_os2.h>
+#include <openssl/crypto.h>
+
+#if defined(OPENSSL_THREADS)
+
+# include <internal/thread.h>
+# include <internal/thread_arch.h>
+# include "openssl_threads.h"
+# include "worker.h"
+
+static int thread_start_now(OSSL_LIB_CTX *, CRYPTO_TASK *);
+static ossl_inline void thread_schedule(OSSL_LIB_CTX_THREADS *, CRYPTO_TASK *);
+static ossl_inline int thread_can_spawn(OSSL_LIB_CTX_THREADS *);
+
+void *crypto_thread_start(OSSL_LIB_CTX *ctx, CRYPTO_THREAD_ROUTINE start,
+                          void *data, uint8_t options)
+{
+    CRYPTO_TASK *task;
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (tdata == NULL || tdata->enabled == 0)
+        return NULL;
+
+    if ((task = crypto_task_new(start, data)) == NULL)
+        return NULL;
+
+    return crypto_thread_start_task(ctx, task, options);
+}
+
+void *crypto_thread_start_task(OSSL_LIB_CTX *ctx, CRYPTO_TASK *task,
+                               uint8_t options)
+{
+    int can_spawn = 0;
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (tdata == NULL || task == NULL)
+        return NULL;
+
+    CRYPTO_MUTEX_lock(tdata->workers.start_mutex);
+
+    CRYPTO_MUTEX_lock(tdata->lock);
+    task->id = tdata->tasks.created_tasks++;
+    list_add_tail(&task->list, &tdata->tasks.queue);
+
+    can_spawn = thread_can_spawn(tdata);
+    CRYPTO_MUTEX_unlock(tdata->lock);
+
+    if (can_spawn == 0) {
+        thread_schedule(tdata, task);
+    }
+    else if (thread_start_now(ctx, task) == 0) {
+        thread_schedule(tdata, task);
+    }
+
+    CRYPTO_MUTEX_lock(task->lock);
+    while (task->state < CRYPTO_TASK_STARTED)
+        CRYPTO_CONDVAR_wait(task->cond, task->lock);
+    CRYPTO_MUTEX_unlock(task->lock);
+
+    CRYPTO_MUTEX_unlock(tdata->workers.start_mutex);
+
+    return (void *) task;
```

Comment:
Nit: remove space between cast and value.

Reply:
Resolved here: https://github.com/openssl/openssl/pull/12255/commits/376cd62b4bc085539c39e38c089d6743d6b03261.